remoting-agent/src/main/java/org/wso2/testgrid/agent/AgentStreamObserver.java

```diff
@@ -0,0 +1,155 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.testgrid.agent;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.testgrid.agent.listeners.OperationResponseListener;
+import org.wso2.testgrid.common.agentoperation.AgentObservable;
+import org.wso2.testgrid.common.agentoperation.Operation;
+import org.wso2.testgrid.common.agentoperation.OperationSegment;
+
+import java.util.Calendar;
+import java.util.Observable;
+import java.util.Observer;
+
+/**
+ * Observer class to wait till response and send response as segment to the
+ * Tinkerer.
+ */
+public class AgentStreamObserver implements Observer {
+
+    private static final int MAX_CONTENT_SIZE = 1024;
+    private static final long MAX_EXECUTION_TIME_OUT = 500;
+    private static final Logger logger = LoggerFactory.getLogger(AgentStreamObserver.class);
+
+    private String operationId;
+    private OperationResponseListener operationResponseListener;
+    private Process process;
+    private long resultSize;
+    private AgentObservable agentObservable;
+    private volatile boolean oneProcessCompleted = false;
+    private volatile long initTime;
+    private volatile String shellLog = "";
+    private volatile boolean abortExecution = false;
+
+    /**
+     * Initialize object constructor.
+     *
+     * @param operationResponseListener         Send response back to the Tinkerer
+     * @param operationId                       The operation id
+     * @param process                           Command execution process
+     * @param agentObservable                   Observable agent
+     */
+    public AgentStreamObserver(OperationResponseListener operationResponseListener, String operationId,
+                               Process process, AgentObservable agentObservable) {
+        this.operationResponseListener = operationResponseListener;
+        this.operationId = operationId;
+        this.resultSize = 0;
+        this.initTime = Calendar.getInstance().getTimeInMillis();
+        this.process = process;
+        this.agentObservable = agentObservable;
+    }
+
+    /**
+     * Handle response from the command execution thread and send response as segment
+     * To the tinkerer.
+     *
+     * @param o                 Observable object
+     * @param arg               Shell execution result
+     */
+    @Override
+    public synchronized void update(Observable o, Object arg) {
+        StreamResponse streamResponse = (StreamResponse) arg;
+        if (!streamResponse.isCompleted()) {
+            this.resultSize += streamResponse.getResponse().length();
+        }
+        if (streamResponse.isCompleted()) {
+            // Response send back after input stream read complete
+            if (this.oneProcessCompleted) {
+                OperationSegment finalOperationSegment = new OperationSegment();
+                finalOperationSegment.setOperationId(this.operationId);
```

Comment:
You could do the initialization of OperationSegment once and set an operationId and use it after.